include/verilated_random.cpp

```diff
@@ -0,0 +1,422 @@
+        return false;
+    }
+    if (obits <= VL_BYTESIZE) {
+        CData* const p = static_cast<CData*>(datap());
+        *p = VL_CLEAN_II(obits, obits, owp[0]);
+    } else if (obits <= VL_SHORTSIZE) {
+        SData* const p = static_cast<SData*>(datap());
+        *p = VL_CLEAN_II(obits, obits, owp[0]);
+    } else if (obits <= VL_IDATASIZE) {
+        IData* const p = static_cast<IData*>(datap());
+        *p = VL_CLEAN_II(obits, obits, owp[0]);
+    } else if (obits <= VL_QUADSIZE) {
+        QData* const p = static_cast<QData*>(datap());
+        *p = VL_CLEAN_QQ(obits, obits, VL_SET_QW(owp));
+    } else {
+        _vl_clean_inplace_w(obits, owp);
+    }
+    return true;
+}
+
+std::shared_ptr<const VlRandomExpr> VlRandomizer::random_constraint(VlRNG& rngr, int bits) {
+    unsigned long long hash = VL_RANDOM_RNG_I(rngr) & ((1 << bits) - 1);
+    std::shared_ptr<const VlRandomExpr> concat = nullptr;
+    std::vector<std::shared_ptr<const VlRandomExpr>> varbits;
+    for (const auto& var : m_vars) {
+        for (int i = 0; i < var.second->width(); i++)
+            varbits.emplace_back(std::make_shared<const VlRandomExtract>(var.second, i));
+    }
+    for (int i = 0; i < bits; i++) {
+        std::shared_ptr<const VlRandomExpr> bit = nullptr;
+        for (unsigned j = 0; j * 2 < varbits.size(); j++) {
+            unsigned idx = j + VL_RANDOM_RNG_I(rngr) % (varbits.size() - j);
+            auto sel = varbits[idx];
+            std::swap(varbits[idx], varbits[j]);
+            bit = bit == nullptr ? sel : std::make_shared<const VlRandomBinOp>("bvxor", bit, sel);
+        }
+        concat = concat == nullptr ? bit
+                                   : std::make_shared<const VlRandomBinOp>("concat", concat, bit);
+    }
+    return std::make_shared<const VlRandomBinOp>(
+        "=", concat, std::make_shared<const VlRandomConst>(hash, bits));
+}
+
+bool VlRandomizer::next(VlRNG& rngr) {
+    if (m_vars.empty()) return true;
+    std::iostream& f = get_solver();
+    if (!f) return false;
+
+    f << "(set-option :produce-models true)\n";
+    f << "(set-logic QF_BV)\n";
+    for (const auto& var : m_vars) {
+        f << "(declare-fun " << var.second->name() << " () (_ BitVec " << var.second->width()
+          << "))\n";
+    }
+    for (const std::string& constraint : m_constraints) { f << "(assert " << constraint << ")\n"; }
+    f << "(check-sat)\n";
+
+    int rc = parse_solution(f);
+    if (rc == 1) {
+        // unsat
+        f << "(reset)\n";
+        return false;
+    }
+    if (rc == 2) {
+        // error
+        f << "(reset)\n";
+        return false;
+    }
+
+    for (int i = 0; i < HASH_LEN_TOTAL && rc == 0; i++) {
+        f << "(assert ";
+        random_constraint(rngr, HASH_LEN)->emit(f);
+        f << ")\n";
+        f << "\n(check-sat)\n";
+        rc = parse_solution(f);
+    }
+
+    f << "(reset)\n";
+    return true;
+}
+
+int VlRandomizer::parse_solution(std::iostream& f) {
```

Comment:
```suggestion
int VlRandomizer::parseSolution(std::iostream& f) {
```

Also add comment what the return value is, as not clear to me (why "2")

Reply:
Renamed; reworked this to use a clearer error path and only return true on sat, false on unsat and parse error, printing warnings.